src/utils/value.rs

```diff
@@ -0,0 +1,261 @@
+//
+macro_rules! impl_value_is {
+    ($method:ident) => {
+        impl_value_is!(
+            $method,
+            Json => $method,
+            Yaml => $method
+        );
+    };
+    ($method:ident, $( $v:ident => $vm:ident ),+ ) => {
+        impl<'a> Value<'a> {
+            fn $method(&self) -> bool {
+                match *self {
+                    $(
+                        Value::$v(x) => x.$vm(),
+                    )+
+                }
+            }
+        }
+    };
+}
+
+macro_rules! impl_value_as {
+    ($method:ident, $result:ty) => {
+        impl_value_as!(
+            $method, $result,
+            Json => $method,
+            Yaml => $method
+        );
+    };
+    ($method:ident, $result:ty, $( $v:ident => $vm:ident ),+ ) => {
+        impl<'a> Value<'a> {
+            fn $method(&self) -> $result {
+                match self {
+                    $(
+                        Value::$v(x) => x.$vm(),
+                    )+
+                }
+            }
+        }
+    };
+}
+
+impl_value_is!(is_bool, Json => is_boolean, Yaml => is_bool);
+impl_value_as!(as_bool, Option<bool>);
+impl_value_is!(is_null);
+impl_value_as!(as_null, Option<()>);
+impl_value_is!(is_i64);
+impl_value_as!(as_i64, Option<i64>);
+impl_value_is!(is_u64);
+impl_value_as!(as_u64, Option<u64>);
+impl_value_is!(is_f64);
+impl_value_as!(as_f64, Option<f64>);
+impl_value_is!(is_string);
+impl_value_as!(as_str, Option<&str>);
+impl_value_is!(is_array, Json => is_array, Yaml => is_sequence);
+impl_value_is!(is_object, Json => is_object, Yaml => is_mapping);
+
+impl<'a> Value<'a> {
+    fn as_array(&'a self) -> Option<Vec<Value<'a>>> {
+        match self {
+            Value::Yaml(yaml) => yaml.as_sequence().map(|x| x.iter().map(Value::from_ref).collect()),
+            Value::Json(json) => json.as_array().map(|x| x.iter().map(Value::from_ref).collect()),
+        }
+    }
+
+    fn as_object(&'a self) -> Option<Result<HashMap<&'a str, Value<'a>>, ()>> {
+        match self {
+            Value::Yaml(yaml) => yaml.as_mapping().map(|x| {
+                x.iter()
+                    .map(|(k, v)| k.as_str().map(|k| (k, Value::from_ref(v))).ok_or(()))
+                    .collect()
+            }),
+            Value::Json(json) => json
+                .as_object()
+                .map(|x| x.iter().map(|(k, v)| Ok((k.as_ref(), Value::from_ref(v)))).collect()),
+        }
+    }
+}
+
+//
+// Comparison
+//
+impl<'a, 'b> PartialEq<Value<'a>> for Value<'b> {
+    fn eq(&self, other: &Value) -> bool {
+        if self.is_null() {
+            return self.as_null() == other.as_null();
+        }
+        if self.is_bool() {
+            return self.as_bool() == other.as_bool();
+        }
+        if self.is_u64() {
+            return self.as_u64() == other.as_u64();
+        }
+        if self.is_i64() {
+            return self.as_i64() == other.as_i64();
+        }
+        if self.is_f64() {
+            // TODO approx crate
```

Comment:
this may be worth writing some place outside of the code, like an issue or something so we don't forget, as this can lead to hard to debug problems later.

Reply:
Agree, but with one exception. Till 1.0 I consider it as okay. Because 1.0 MUST not contain any `TODO` comments. They should be replaced with issues if they won't be fixed.